src/rdkafka_int.h

```diff
@@ -912,7 +918,46 @@ void rd_kafka_log0(const rd_kafka_conf_t *conf,
                 }                                              \
         } while (0)
 
-
+/*
+        Wait for available memory space before allocating the expected memory size.
+        We use compare-and-swap (CAS) to register the memory one thread wants to allocate. CAS only
+        succeed if the total memory usage is still the same when checking. If another thread
+        already register its expected memory usage, the current thread needs check for the total
+        memory usage again.
+*/
+#define rd_wait_mem_space(expected_size, succeeded)  do {                                               \
```

Comment:
Why do it as a macro? `succeeded` could be returned from a function call and compiler could inline it?

Reply:
This is what I am thinking too. Originally, I think it was a small piece of code. Now we can extract it as a function.